### PR TITLE
routing: handle failure to launch shard after permanent failure

### PR DESCRIPTION
**Pull request by carlaKC**
As per @halseth's [exceptional explanation](https://github.com/lightningnetwork/lnd/issues/4788#issuecomment-804093718), our payment lifecycle fails to cover the following scenario:
- Shard 0 is dispatched
- Shard 1 is dispatched
- Path finding for shard 2 begins
- Shard 0 returns a fatal error
- Shard 2's [`launchShard`](https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L249) fails with `ErrPaymentAlreadyFailed`
- Shard 1 gets stuck in flight and the payment's state is never updated to failed

Fixes: #4788
Fixes: #4501

The [handling](https://github.com/lightningnetwork/lnd/blob/a329c8061266418bccd797aa5d7d277c736ee930/routing/router.go#L1911) in `SendToRoute` is not updated because we want to fail attempts to launch shards for already failed payments, and we don't have the problem of not following up with the other shards.

### CODE REVIEW

File: routing/mock_test.go [hunk 1]
```diff
@@ -83,15 +83,19 @@ func (m *mockPaymentAttemptDispatcher) setPaymentResult(
 }
 
 type mockPaymentSessionSource struct {
-	routes []*route.Route
+	routes       []*route.Route
```
Comment (Roasbeef) [inline routing/mock_test.go:86]: typo in commit body: onc -> one

File: routing/payment_lifecycle_test.go [hunk 2]
```diff
@@ -575,7 +611,7 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 				paymentError,
 			},
 			routes: []*route.Route{
-				shard, shard, shard, shard, shard, shard,
+				shard, shard, shard, shard,
```
Comment (Roasbeef) [inline routing/payment_lifecycle_test.go:626]: Why do the last two shards get eliminated?

File: routing/payment_lifecycle_test.go [hunk 3]
```diff
@@ -627,6 +627,40 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 			},
 			paymentErr: channeldb.FailureReasonPaymentDetails,
 		},
+		{
+			// A MP payment scenario when our path finding returns
+			// after we've just received a terminal failure, and
```
Comment (Roasbeef) [inline routing/payment_lifecycle_test.go:632]: Excellent bit of refactoring of the tests up to this point leading o the addition of this new reproduction test case!

File: routing/payment_lifecycle.go [hunk 4]
```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 
 		// We found a route to try, launch a new shard.
 		attempt, outcome, err := shardHandler.launchShard(rt)
-		if err != nil {
+		switch err {
+		// We can get an already failed error in the case where we
```
Comment (Roasbeef) [inline routing/payment_lifecycle.go:251]: So now we'll go back to the top, and exit as we now have a failure (`state.terminate==True`) and there're no more active shards?

**Review by Roasbeef — COMMENTED** [review]
Great job with this fix! Also shouts out to @halseth for the crisp description of the root cause. Completed an initial take of the PR, but will need to circle back once I familiarize myself a bit more with the set of tests being modified to lead to the repro test case itself.

File: routing/payment_lifecycle_test.go [hunk 5]
```diff
@@ -48,6 +51,110 @@ func createTestRoute(amt lnwire.MilliSatoshi,
 	)
 }
 
+// paymentLifecycleTestCase contains the steps that we expect for a payment
+// lifecycle test, and the routes that pathfinding should deliver.
+type paymentLifecycleTestCase struct {
```
Comment (yyforyongyu) [inline routing/payment_lifecycle_test.go:56]: Nice refactor! I'm actually working on this file too, guess I'll rebase on your commit haha.

File: routing/payment_lifecycle_test.go [hunk 6]
```diff
@@ -502,33 +565,44 @@ func TestRouterPaymentStateMachine(t *testing.T) {
...
 				routerSettleAttempt,
 				paymentSuccess,
 			},
-			routes: []*route.Route{shard, shard, shard, shard},
+			routes: []*route.Route{shard, shard, shard, rt},
 		},
 		{
 			// An MP payment scenario a shard fail with a terminal
 			// error, causing the router to stop attempting.
+			name: "MP terminal",
```
Comment (yyforyongyu) [inline routing/payment_lifecycle_test.go:584]: When I ran this test multiple times, sometime I would get a fail. Curious if it ever happen to you? 
```
    // The first shard fail with a terminal error.
    getPaymentResultTerminalFailure,
    routerFailAttempt,
    routerFailPayment,
```
After this setup, the shardHandler will have a terminal state and it will shut down, causing the following steps failed,
```
    // Remaining 3 shards fail.
    getPaymentResultTempFailure,
    getPaymentResultTempFailure,
    getPaymentResultTempFailure,
    routerFailAttempt,
    routerFailAttempt,
    routerFailAttempt,
```

File: routing/payment_lifecycle_test.go (hunk 3 above)
Comment (halseth) [reply routing/payment_lifecycle_test.go:632]: +1

File: routing/payment_lifecycle.go (hunk 4 above)
Comment (carlaKC) [reply routing/payment_lifecycle.go:251]: Yep exactly, we're def in a terminate state so now we just need to wait for our shards to come back (if any). 

Wondering whether we should `waitForShard` for each of our `numActiveShards` here? Otherwise we'll loop through checkshards/ (pointless) pathfinding / err already failed while we wait for remaining shards to come back and we hit `state.terminate && state.numShardsInFlight == 0`.

File: routing/payment_lifecycle_test.go (hunk 2 above)
Comment (carlaKC) [reply routing/payment_lifecycle_test.go:626]: This test actually never uses them, it just needs 4 shards then it succeeds. Pulled this out into a separate commit so it's a bit clearer.

File: routing/payment_lifecycle.go [hunk 7]
```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
...
+		// received a permanent failure while we were pathfinding for
+		// this shard.
+		case channeldb.ErrPaymentAlreadyFailed:
```
Comment (halseth) [inline routing/payment_lifecycle.go:254]: I initially found it a bit weird to handle the error here and where the control tower is actually called. But looking at it more closely this might be best, since otherwise we would have to signal back to this loop whether to expect a result to be collected or not (re-purposing the `launchOutcome` would be an option).

Just saying there might exist a cleaner way to do this, but I'm sure this is correct :)

File: routing/payment_lifecycle.go [hunk 8]
```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
...
+			log.Infof("Payment: %v failed, abandoning shard",
+				p.paymentHash)
+
+			continue
+
+		case nil:
```
Comment (halseth) [inline routing/payment_lifecycle.go:260]: style suggesion:
```
swtich {
case ErrPaymentAlreadyFailed:
   ... continue
case err != nil:
     return
 }
 ```

**Review by halseth — CHANGES_REQUESTED** [review]
Great solution 👍

File: routing/payment_lifecycle_test.go (hunk 6 above)
Comment (carlaKC) [reply routing/payment_lifecycle_test.go:584]: Did this happen in this commit specifically? I think that this test case also hits the bug that is fixed in the following commits, which may be causing the failure. Found a few other races in these tests this morning, so will push a new version with fixes soon.

File: routing/payment_lifecycle_test.go [hunk 9]
```diff
@@ -526,13 +553,18 @@ func TestRouterPaymentStateMachine(t *testing.T) {
...
+				sendToSwitchSuccess,
+
 				// The fourth shard succeed against all odds,
 				// making the overall payment succeed.
 				getPaymentResultSuccess,
 				routerSettleAttempt,
 				paymentSuccess,
 			},
-			routes: []*route.Route{shard, shard, shard, shard},
+			routes: []*route.Route{shard, shard, shard, rt},
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/payment_lifecycle_test.go:579]: I was misunderstanding this test case's purpose, reverted 👍

File: routing/payment_lifecycle.go (hunk 7 above)
Comment (carlaKC) [reply routing/payment_lifecycle.go:254]: Also considered this approach, ended up leaving it in here because this case only affects this call to `launchShard` because we're operating in the context of this loop and (I think?) we want to leave `SendToRoute` as is and error out on `ErrPaymentAlreadyFailed`?

**Comment by carlaKC** [comment]
Pushed a few changes to how the tests are refactored, the reason for some changes are difficult to string together so here's a summary:

To reproduce this bug we need the following changes in the tests:
1. Block on `RequestRoute`: otherwise we can't reliably hit the condition where we launch a shard after failure is processed
2. Assert `paymentError` value: this helps to demonstrate why we're failed our payment

To add these two features to the tests, I needed to alter existing tests to prevent races:

**`MP one shard success`**: this tests the case where we learn the preimage for a payment, even if we didn't pay the full amount.
- Problem: while we're waiting for our single shard to succeed, the payment loop will continue to try to dispatch shards for the payment, because we haven't sent our full amount yet. This can happen 1 or two times depending on how quickly `collectAsync` processes the success shard. We don't know how many times to unblock the router, so the test flakes.
- [Change](https://github.com/lightningnetwork/lnd/pull/5214/commits/b3f02d6d925b01ce18c915645a3e7368f2e66a20): increase the shard size to half the payment so that the test will reliably look for a single shard while we wait on our success shard.

**`single shot restart`**: this tests the case where we shutdown while processing a payment.
- Problem: when we restart a payment in our test, we sent an [arbitrary error back from the switch](https://github.com/lightningnetwork/lnd/blob/a329c8061266418bccd797aa5d7d277c736ee930/routing/payment_lifecycle_test.go#L824). This does not match our actual behaviour of [closing the result channel.](https://github.com/lightningnetwork/lnd/blob/a329c8061266418bccd797aa5d7d277c736ee930/htlcswitch/switch.go#L419) The test can flake depending on whether we receive this mocked error from  `collectShardAsyc` or `ErrRouterShuttingDown` (checked in many places in the main loop).
- [Change](https://github.com/lightningnetwork/lnd/pull/5214/commits/46aecf8312616ef93dc443ad24ca3976c7aa1925): update mock to close the channel when we restart. This better mimics the switch + we don't get an error from `collectShardAsync` anymore, so we reliably fail with `ErrRouterShuttingDown`

File: routing/payment_lifecycle_test.go [hunk 10]
```diff
@@ -526,13 +553,18 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 				routerFailAttempt,
 				routerFailAttempt,
 
+				// shard 3
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/payment_lifecycle_test.go]: Put this back to dispatching all shards first, but did make some other changes - explained [here](https://github.com/lightningnetwork/lnd/pull/5214#issuecomment-821067208). I think the change is ok because it still covers the "settles less than total amount" case?

File: routing/payment_lifecycle_test.go [hunk 11]
```diff
@@ -617,6 +617,40 @@ func TestRouterPaymentStateMachine(t *testing.T) {
...
+				sendToSwitchSuccess,
+
+				// The first shard fail with a terminal error.
+				getPaymentResultTerminalFailure,
+				routerFailAttempt,
+				routerFailPayment,
+
+				// shard 1
+				routeRelease,
+				routerRegisterAttempt,
```
Comment (halseth) [inline routing/payment_lifecycle_test.go:644]: should add a comment here that we expect it to register the attempt, but this actually fails. Do we have a way of assersting here that the registration fails with the `ErrPaymentAlreadyFailed` error?

File: routing/payment_lifecycle.go [hunk 12]
```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
...
+		// We can get an already failed error in the case where we
+		// received a permanent failure while we were pathfinding for
+		// this shard. We know that we are in a terminal state, because
+		// we've received this error, so we can continue and wait for
+		// our remaining shards to complete.
+		case err == channeldb.ErrPaymentAlreadyFailed:
+			log.Infof("Payment: %v failed, abandoning shard",
+				p.paymentHash)
+
+			continue
```
Comment (halseth) [inline routing/payment_lifecycle.go:260]: Can label the loop to make this a bit easier to follow perhaps.

File: routing/payment_lifecycle.go (hunk 4 above)
Comment (halseth) [reply routing/payment_lifecycle.go:251]: won't we go into the second case `state.terminate || state.remainingAmt == 0` which will cal `waitForShards`?

**Review by halseth — APPROVED** [review]
Great work! A few minor comments from me, LGTM 👍

File: routing/payment_lifecycle_test.go (hunk 11 above)
Comment (calvinrzachman) [reply routing/payment_lifecycle_test.go:644]: This sounds like a good idea to me. I think this solution will cover us in the test, but I have a question as to whether this will cover this bug case "in the wild".

The definition of a "FAILED" payment looks a bit different in testing as compared to the live `resumePayment()` payment lifecycle machine. See the following:
- [Register Attempt (Mock)](https://github.com/lightningnetwork/lnd/blob/master/routing/mock_test.go#L294-L296)
- [Register Attempt (Live)](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294)

Both check to see whether the top level payment is "FAILED" before allowing a new shard registration. The mock determines failure with a simple lookup in the `mockControlTower`'s `failed` map. However in the live case, when we [register a new shard with the ControlTower/DB](https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L380), we fetch the payment and [ensure the payment is `IN-FLIGHT`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294). With the way [`fetchPayment()`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payments.go#L358-L365) works, a payment is reported as `status=IN-FLIGHT` by default if it has any in-flight shards. Given this, I don’t see how we would be returning the [`ErrPaymentAlreadyFailed`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L670). If [`shard 1`](https://github.com/lightningnetwork/lnd/pull/5214#issue-615993831) is stuck `IN-FLIGHT`, it looks like `fetchPayment()` will tell us the payment is still in flight.

I think we might be hitting the [terminal info check](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L300-L302) just below (which exists in the wild but not in our mock). This is a looser check which is sensitive to any payment level failure and does not care about whether there are in flight shards or not. In this case, it would not be `ErrPaymentAlreadyFailed`, but actually `ErrPaymentTerminal` which we're hitting in the wild.

Here’s a diff with what I think will get the mock of `RegisterAttempt` to be more inline with how it looks like it works in the wild: https://github.com/lightningnetwork/lnd/commit/4e388c00571d4aa69b61ff42f978016941dcfa91

File: routing/payment_lifecycle.go [hunk 13]
```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
...
 		// We found a route to try, launch a new shard.
 		attempt, outcome, err := shardHandler.launchShard(rt)
-		if err != nil {
+		switch {
+		// We can get an already failed error in the case where we
+		// received a permanent failure while we were pathfinding for
+		// this shard. We know that we are in a terminal state, because
+		// we've received this error, so we can continue and wait for
+		// our remaining shards to complete.
+		case err == channeldb.ErrPaymentAlreadyFailed:
```
Comment (calvinrzachman) [inline routing/payment_lifecycle.go:256]: While reviewing I noticed a difference between test vs. production code (pre-dates this PR), which looks like it may have an impact on behavior. I think this solution will cover us in the test, but I have a question as to whether this will cover this bug case "in the wild".

The definition of a "FAILED" payment looks a bit different in testing as compared to the live `resumePayment()` payment lifecycle machine. See the following:
- [Register Attempt (Mock)](https://github.com/lightningnetwork/lnd/blob/master/routing/mock_test.go#L294-L296)
- [Register Attempt (Live)](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294)

Both check to see whether the top level payment is "FAILED" before allowing a new shard registration. The mock determines failure with a simple lookup in the `mockControlTower`'s `failed` map. However in the live case, when we [register a new shard with the ControlTower/DB](https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L380), we fetch the payment and [ensure the payment is `IN-FLIGHT`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294). With the way [`fetchPayment()`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payments.go#L358-L365) works, a payment is reported as `status=IN-FLIGHT` by default if it has _*any in-flight shards*_. Given this, I'm not certain that we would be returning [`ErrPaymentAlreadyFailed`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L670). If [`shard 1`](https://github.com/lightningnetwork/lnd/pull/5214#issue-615993831) is stuck `IN-FLIGHT`, it looks like `fetchPayment()` will tell us the payment is still in flight.

I think we might be hitting the [terminal info check](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L300-L302) just below (which exists in the wild but I can't find in our mock). This is a looser check which is sensitive to any payment level failure, but does not care about whether there are in flight shards or not. In this case, it would not be `ErrPaymentAlreadyFailed`, but actually `ErrPaymentTerminal` which we're hitting in the wild.

Let me know if this explanation is difficult to grok or if I'm making up phantom differences which don't actually exist. Here’s a diff with what I think will get the mock of `RegisterAttempt` to be more inline with how it looks like it works in the wild: https://github.com/lightningnetwork/lnd/commit/cb22a03850a78bbfd630c3aa3f4a9d4bb6633971

Comment (halseth) [reply routing/payment_lifecycle.go:256]: I agree, we should do the terminal info check first, that's the most important condition that must be satisfied.

Good catch on the mocks 👍

File: channeldb/payment_control.go [hunk 14]
```diff
@@ -290,18 +290,19 @@ func (p *PaymentControl) RegisterAttempt(paymentHash lntypes.Hash,
...
 
-		// Ensure the payment is in-flight.
-		if err := ensureInFlight(p); err != nil {
-			return err
-		}
-
 		// We cannot register a new attempt if the payment already has
-		// reached a terminal condition:
+		// reached a terminal condition. We check this before
+		// ensureInFlight because it is a more general check.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply channeldb/payment_control.go:295]: From [this line](https://github.com/lightningnetwork/lnd/blob/master/channeldb/payments.go#L353), suppose the payment has two HTLC attempts, `h1`, `h2`, then,

`h1` settled, and

- `h2` settled, then we have, `inflight=false && settled=true`, go to `case !inflight&& settled` => `StatusSucceeded`
- `h2` failed => then we have, `inflight=false && settle=true`, go to `case !inflight && settled` => `StatusSuccedded`
- `h2` is neither settled or failed, then we have, `inflight=true && settle=true`, go to `default` => `StatusInFlight`

`h1` failed, and

- `h2` settled, then we have `inflight=false && settled=true`, go to `case !inflight && settled` => `StatusSucceeded`
- `h2` failed, then we have `inflight=false && settled=false`, then,
  - if `failureReason != nil` , => `StatusFailed`
  - otherwise, `StatusInFlight`
- `h2` is neither settled or failed, then we have `inflight=true && settled=false`, => `StatusInFlight`

`h1` is neither settled or failed

- `h2` settled, then we have `inflight=true && settled=true`, go to `default` => `StatusInFlight`
- `h2` failed, then we have `inflight=true && settled=false`, go to `default` => `StatusInFlight`
- `h2` is neither settled or failed, then we have `inflight=true && settled=false`, go to `default` => `StatusInFlight`

[the remark this answers is not in the document]
Comment (yyforyongyu) [reply channeldb/payment_control.go:295]: My question is, why are we relying on the payment's `failureReason`but not the `HTLCFailInfo.Failre` here?

File: routing/payment_lifecycle_test.go (hunk 6 above)
Comment (yyforyongyu) [reply routing/payment_lifecycle_test.go:574]: Nope. This happens in another PR that I made, after refactoring the `shardHandler`. I think you got it fixed here. The problem was multiple shards were running, one of them got a terminal error causing it to shut down, depending on the timing, it may cause error in other shards.

File: routing/mock_test.go [hunk 15]
```diff
@@ -279,13 +279,13 @@ func (m *mockControlTower) InitPayment(phash lntypes.Hash,
...
 	a *channeldb.HTLCAttemptInfo) error {
 
-	m.Lock()
-	defer m.Unlock()
-
 	if m.registerAttempt != nil {
 		m.registerAttempt <- registerAttemptArgs{a}
 	}
 
+	m.Lock()
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/mock_test.go:302]: If we have nil channels in the mock, I believe it means that we just don't want to block on these calls, but we still want to lock to be thread-safe.

File: routing/payment_lifecycle_test.go [hunk 16]
```diff
@@ -871,11 +882,12 @@ func testPaymentLifecycle(t *testing.T, test paymentLifecycleTestCase,
 		// In this state we expect to receive an error for the
 		// original payment made.
 		case paymentError:
+			require.NotNil(t, test.paymentErr,
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/payment_lifecycle_test.go:1010]: Always forget about this, thanks!

File: routing/mock_test.go [hunk 17]
```diff
@@ -302,20 +302,38 @@ func (m *mockControlTower) RegisterAttempt(phash lntypes.Hash,
...
 	defer m.Unlock()
 
+	// Lookup payment.
+	p, ok := m.payments[phash]
+	if !ok {
+		return channeldb.ErrPaymentNotInitiated
+	}
+
+	var inFlight bool
+	for _, a := range p.attempts {
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/mock_test.go:312]: Always a tradeoff there, because we could be running tests that mock `RegisterAttempt` calls that return the wrong value - part of why we never caught this bug in the first place. I think that it could actually be worth using a test db in these because there's a lot of state logic in some of the payment control calls? But perhaps in a follow up since there's a lot of refactor-tractoring in this already.

File: channeldb/payment_control.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (halseth) [reply channeldb/payment_control.go:295]: Yep, you got it right 👌 

> why are we relying on the payment's failureReason but not the HTLCFailInfo.Failre here?

Not entirely sure what you mean, but the `failureReason` holds the overall payment failure state, while a HTLC failure doesn't really impact the payment failure state (since we will just retry with new HTLCs). The only exception is when an HTLC failure indicates that we should terminate the payment, in which case we must wait for no HTLCs to be in flight before we can determine the final state of the payment.

File: routing/payment_lifecycle_test.go [hunk 18]
```diff
@@ -598,12 +598,12 @@ func testPaymentLifecycle(t *testing.T, test paymentLifecycleTestCase,
 	// Create a mock control tower with channels set up, that we use to
 	// synchronize and listen for events.
 	control := makeMockControlTower()
-	control.init = make(chan initArgs, 20)
-	control.registerAttempt = make(chan registerAttemptArgs, 20)
-	control.settleAttempt = make(chan settleAttemptArgs, 20)
-	control.failAttempt = make(chan failAttemptArgs, 20)
-	control.failPayment = make(chan failPaymentArgs, 20)
-	control.fetchInFlight = make(chan struct{}, 20)
+	control.init = make(chan initArgs)
```
Comment (halseth) [inline routing/payment_lifecycle_test.go:723]: will probably help with some flakes 😅

File: routing/mock_test.go [hunk 19]
```diff
@@ -248,13 +248,13 @@ func makeMockControlTower() *mockControlTower {
 func (m *mockControlTower) InitPayment(phash lntypes.Hash,
 	c *channeldb.PaymentCreationInfo) error {
 
-	m.Lock()
```
Comment (halseth) [inline routing/mock_test.go:251]: great commit message! 👍

File: routing/payment_lifecycle_test.go [hunk 20]
```diff
@@ -190,6 +190,9 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 		t.Fatalf("unable to create route: %v", err)
 	}
 
+	halfShard, err := createTestRoute(paymentAmt/2, testGraph.aliasMap)
```
Comment (halseth) [inline routing/payment_lifecycle_test.go:197]: Once again, excellent commit message 🙏  🙇

File: routing/payment_lifecycle_test.go [hunk 21]
```diff
@@ -698,7 +698,7 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 			routes: []*route.Route{
 				shard, shard, shard, shard,
 			},
-			paymentErr: channeldb.ErrPaymentTerminal,
+			paymentErr: channeldb.FailureReasonPaymentDetails,
```
Comment (halseth) [inline routing/payment_lifecycle_test.go:701]: Great 👍  I was wondering why the `paymentErr` was `ErrPaymentTerminal` in the commit this test was added, but I guess that was just to make the test pass before you fixed the logic?

File: routing/mock_test.go [hunk 22]
```diff
@@ -325,6 +325,10 @@ func (m *mockControlTower) RegisterAttempt(phash lntypes.Hash,
 	_, settled := m.successful[phash]
 	_, failed := m.failed[phash]
 
+	if settled || failed {
+		return channeldb.ErrPaymentTerminal
+	}
+
```
Comment (halseth) [inline routing/mock_test.go:331]: I guess below conditions now can never be hit

File: routing/payment_lifecycle_test.go [hunk 23]
```diff
@@ -619,7 +619,7 @@ func TestRouterPaymentStateMachine(t *testing.T) {
 			// A MP payment scenario when our path finding returns
 			// after we've just received a terminal failure, and
 			// demonstrates a bug where the payment will return with
-			// and "unexpected" ErrPaymentAlreadyFailed rather than
+			// and "unexpected" ErrPaymentTerminal rather than
```
Comment (halseth) [inline routing/payment_lifecycle_test.go:622]: doesn't demonstrate a bug anymore 😁

File: routing/payment_lifecycle.go [hunk 24]
```diff
@@ -247,7 +247,18 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 
 		// We found a route to try, launch a new shard.
 		attempt, outcome, err := shardHandler.launchShard(rt)
-		if err != nil {
+		switch {
```
Comment (halseth) [inline routing/payment_lifecycle.go:250]: Not sure why I cannot comment on it, but did we resolve this convo? https://github.com/lightningnetwork/lnd/pull/5214#discussion_r614453436

File: routing/payment_lifecycle.go (hunk 12 above)
Comment (halseth) [reply routing/payment_lifecycle.go:259]: Bump on this one

**Review by halseth — APPROVED** [review]
Excellent work on this one! This is a very complex part of the codebase, so it is easy to get edge cases wrong. Great work on the tests and refactoring, I think that will help us keep it more robust. 🙏 

 This will finally resolve a bug that has been haunting us for a while, excited to have this finally land! 🎉 

LGMT (only a few nits) ✅

**Review by yyforyongyu — APPROVED** [review]
LGTM!  Looking forward to rebase on this one haha

**Review by calvinrzachman — APPROVED** [review]

File: routing/payment_lifecycle.go (hunk 24 above)
Comment (carlaKC) [reply routing/payment_lifecycle.go:250]: Ah yeah sorry, thought I'd replied! We will indeed hit `state.terminate || state.remainingAmt == 0` and then wait for our shards 👍

File: routing/payment_lifecycle_test.go (hunk 21 above)
Comment (carlaKC) [reply routing/payment_lifecycle_test.go:701]: Yeah! Just to make sure we're actually hitting that error, then add the fix & update test.
